src/validatefs/validatefs.c

```diff
@@ -0,0 +1,342 @@
+        if (r < 0) {
+                if (r != -ENODATA && !ERRNO_IS_NOT_SUPPORTED(r))
+                        return log_error_errno(r, "Failed to read 'user.validatefs.gpt_label' xattr: %m");
+        } else if (!utf8_is_valid(f.gpt_label) || string_has_cc(f.gpt_label, /* ok= */ NULL))
+                return log_error_errno(
+                                SYNTHETIC_ERRNO(EINVAL),
+                                "Extended attribute 'user.validatefs.gpt_label' contains invalid characters, refusing.");
+
+        _cleanup_free_ char *nulstr = NULL;
+        size_t nulstr_size = 0;
+        r = getxattr_at_malloc(fd, /* path= */ NULL, "user.validatefs.mount_point", AT_EMPTY_PATH, &nulstr, &nulstr_size);
+        if (r < 0) {
+                if (r != -ENODATA && !ERRNO_IS_NOT_SUPPORTED(r))
+                        return log_error_errno(r, "Failed to read 'user.validatefs.mount_point' xattr: %m");
+        } else {
+                _cleanup_strv_free_ char **l = strv_parse_nulstr(nulstr, nulstr_size);
+                if (!l)
+                        return log_oom();
+
+                STRV_FOREACH(i, l) {
+                        if (!utf8_is_valid(*i) ||
+                            string_has_cc(*i, /* ok= */ NULL) ||
+                            !path_is_absolute(*i) ||
+                            !path_is_normalized(*i))
+                                return log_error_errno(
+                                                SYNTHETIC_ERRNO(EINVAL),
+                                                "Path listed in extended attribute 'user.validatefs.mount_point' is not a valid, normalized, absolute path or contains invalid characters, refusing: %s", *i);
+                }
+
+                f.mount_point = TAKE_PTR(l);
+        }
+
+        r = !sd_id128_is_null(f.gpt_type_uuid) || f.gpt_label || !strv_isempty(f.mount_point);
+        *ret = TAKE_STRUCT(f);
+        return r;
+}
+
+static int validate_fields_check(int fd, const char *path, const ValidateFields *f) {
+        int r;
+
+        assert(fd >= 0);
+        assert(path);
+        assert(f);
+
+        if (!strv_isempty(f->mount_point)) {
+                _cleanup_free_ char *bad = NULL;
+                bool good = false;
+
+                STRV_FOREACH(i, f->mount_point) {
+                        _cleanup_free_ char *jj = NULL;
+                        const char *j;
+
+                        if (arg_root) {
+                                jj = path_join(arg_root, *i);
+                                if (!jj)
+                                        return log_oom();
+
+                                j = jj;
+                        } else
+                                j = *i;
+
+                        if (path_equal(path, j)) {
+                                good = true;
+                                break;
+                        }
+
+                        if (!strextend_with_separator(&bad, ",", j))
```

Comment:
Not a matter, but the naming `bad` sounds spurious.
We usually use `joined` or so.
Also, maybe better to add a space after comma: `","` -> `", "`

Also, this is just for logging. So, maybe better to use `strv_join()` in `if (!good)` branch.